### PR TITLE
Use temporary oauth2 tokens for Google registries

### LINKED ISSUE #62: How could we authenticate with a service account key with image-syncer?
To give you more context of my problem, we are trying to sync image between the cloud registry and our on-prem registry. For the cloud registry, we need to authenticate with a service account which has no password, only the service account key json. 

How could we authenticate using image syncer for this case? Should we paste the service account key json to the password field? I tried to do so, but seems it is not possible and not a good way.

Kindly seek your help on that. Thank you!

### PR DESCRIPTION
**Pull request by DennisD2**
Pull Request Description
---

### Describe what this PR does / why we need it
For Google registries (*.gcr.io), an OAUTH2 token can be created from a (base64 encoded) service account token.
Define credentials like this:
```
username: _oauth2_
password: <base64 encoded permanent service account token>
```

### Does this pull request fix one issue?

Fixes #62 

### Describe how you did it
Added method for getting the temporary OAUTH2 token from the permanent service account token

### Describe how to verify it
Define a google repo (with gcr.io in hostname), with credentials including a permanent service account token as password value, use the string ```_oauth2_``` as username, and use this as image sync target. The sync will work, and use an on-the-fly created OAUTH2  token

### Special notes for reviews

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,7 +1,9 @@
-image-syncer
```
Comment (hhyasdf) [inline .gitignore:1]: Maybe configuration of `.gitignore` file should not be included to a commit if it seems personal and piecemeal.

And the `image-syncer` binary file below should not be included in a commit either.

File: pkg/sync/destination.go [hunk 2]
```diff
@@ -54,6 +59,19 @@ func NewImageDestination(registry, repository, tag, username, password string, i
 
 	ctx := context.WithValue(context.Background(), ctxKey{"ImageDestination"}, repository)
 	if username != "" && password != "" {
+		fmt.Printf("Credential processing for %s/%s ...\n", registry, repository)
+		if strings.Contains(registry, ".gcr.io") {
```
Comment (hhyasdf) [inline pkg/sync/destination.go:63]: It seems that this commit changes all the behaviors of synchronization about `"*.gcr.io"` registry, which seems dangerous.

So I'm wondering if it's a better way that, we try to get oauth2 token only if the `username` is `"oauth2accesstoken"` and the `registry` string contains `".gcr.io"`. Then the old behaviors will not be affected.

Comment (DennisD2) [reply pkg/sync/destination.go]: Have added code for that. The "username" is checked against "_json_key" which identifies the permament service account token.

File: pkg/sync/destination.go [hunk 3]
```diff
@@ -76,6 +92,11 @@ func NewImageDestination(registry, repository, tag, username, password string, i
 	}, nil
 }
 
+// isPermanentServiceAccountToken returns true if user is a Google permanent service account token
+func isPermanentServiceAccountToken(registry string, username string) bool {
+	return strings.Contains(registry, ".gcr.io") && strings.Compare(username, "_json_key") == 0
```
Comment (hhyasdf) [inline pkg/sync/destination.go:97]: Could you please use a const for the `"_json_key"` here? And maybe `"_json_key"` is not intuitive enough, maybe something like `"_oauth2_"` is better.

File: pkg/sync/destination.go [hunk 4]
```diff
@@ -76,6 +92,11 @@ func NewImageDestination(registry, repository, tag, username, password string, i
 	}, nil
 }
 
+// isPermanentServiceAccountToken returns true if user is a Google permanent service account token
+func isPermanentServiceAccountToken(registry string, username string) bool {
```
Comment (hhyasdf) [inline pkg/sync/destination.go:96]: Maybe this function is better to be put into pkg/sync/utils.go.

Comment (hhyasdf) [inline pkg/sync/destination.go:96]: Maybe this function is better to be put into pkg/sync/utils.go.

File: pkg/sync/destination.go [hunk 5]
```diff
@@ -123,3 +144,22 @@ func (i *ImageDestination) GetRepository() string {
 func (i *ImageDestination) GetTag() string {
 	return i.tag
 }
+
+func gcpTokenFromCreds(creds string) (string, time.Time, error) {
```
Comment (hhyasdf) [inline pkg/sync/destination.go:148]: Maybe this function is better to be put into pkg/sync/utils.go.

Comment (hhyasdf) [inline pkg/sync/destination.go:148]: Maybe this function is better to be put into pkg/sync/utils.go too.

File: pkg/sync/destination.go (hunk 4 above)
Comment (DennisD2) [reply pkg/sync/destination.go:96]: Have moved the function

File: pkg/sync/destination.go (hunk 3 above)
Comment (DennisD2) [reply pkg/sync/destination.go:97]: Done, with "_oauth2_".

File: pkg/sync/destination.go (hunk 5 above)
Comment (DennisD2) [reply pkg/sync/destination.go:142]: Moved it

**Comment by hhyasdf** [comment]
@DennisD2 thanks, we almost done here! Can you fix the golangci-lint error?

**Comment by DennisD2** [comment]
Sorry was some days off. Will check the lint issue now.

**Comment by DennisD2** [comment]
All green now :-)

**Comment by hhyasdf** [comment]
Thanks for your contribution!

**Pull request merged**